app-backend/app/src/main/scala/toolrun/Routes.scala

```diff
@@ -0,0 +1,96 @@
+package com.azavea.rf.toolrun
+
+import java.util.UUID
+
+import scala.util.{Success, Failure}
+import akka.http.scaladsl.model.StatusCodes
+import akka.http.scaladsl.server.Route
+
+import com.azavea.rf.common.{Authentication, UserErrorHandler}
+import com.azavea.rf.database.Database
+import com.azavea.rf.database.tables.ToolRuns
+import com.azavea.rf.datamodel.{PaginatedResponse, ToolRun}
+import com.lonelyplanet.akka.http.extensions.PaginationDirectives
+
+import scala.concurrent.ExecutionContext.Implicits.global
+
+trait ToolRunRoutes extends Authentication
+    with PaginationDirectives
+    with ToolRunQueryParametersDirective
+    with UserErrorHandler {
+  implicit def database: Database
+
+  val toolRunRoutes: Route = handleExceptions(userExceptionHandler) {
+    pathEndOrSingleSlash {
+      get { listToolRuns } ~
+      post { createToolRun }
+    } ~
+    pathPrefix(JavaUUID) { runId =>
+      pathEndOrSingleSlash {
+        get { getToolRun(runId) } ~
+        put { updateToolRun(runId) } ~
+        delete { deleteToolRun(runId) }
+      }
+    }
+  }
+
+  def listToolRuns: Route = authenticate { user =>
+    (withPagination & toolRunQueryParameters) { (page, runParams) =>
+      complete {
+        val toolRunResult = ToolRuns.listToolRuns(page.offset, page.limit, runParams)
+        val numToolRuns = database.db.run(toolRunResult.nRecords)
+        val toolRuns = database.db.run(toolRunResult.records)
+
+        for {
+          toolRunCount <- numToolRuns
+          toolRunPage <- toolRuns
+        } yield {
+          val hasNext = (page.offset + 1) * page.limit < toolRunCount
+          val hasPrevious = page.offset > 0
+          PaginatedResponse[ToolRun](toolRunCount, hasPrevious, hasNext,
```

Comment:
Yeah, I like using this pattern where we can; it's an improvement because it prevents the database code from having to know what the API response should look like. I've updated to use the `ActionRunner` trait here. The only thing that was slightly awkward was that I had to pass `page.offset` and `page.limit` twice, once to the DB and once to `list()`, but it's still a lot cleaner than it was.